### DISCUSSION TITLE
Quais habilidades de um front-end o mercado de trabalho exige hoje em dia?

### DISCUSSION THREAD

**Discussion by LennonSantos** [frontendbr/forum#815] [Discussão]
Senhores, hoje em dia fica evidente a quantidade de tecnologias para se aprender, e me surge a dúvida do que o mercado esta exigindo, exemplo.. o que um dev web precisa saber sobre javascript ou qual tecnologia é essencial. Essa dúvida é pensando para quem quer entrar no mercado ou no meu caso como dev web Jr. algum dia precise procurar alguma oportunidade.

**Comment by matheusml** [comment]
Excelente pergunta @LennonSantos 

Nós aqui na [Concrete](https://concrete.com.br/) fizemos um documento pra guiar os novos colaboradores do Capítulo de JavaScript. Segue o doc abaixo:

# Níveis de carreira de um(a) Front-End Engineer

## Front-End Engineer Básico

Requisitos para Front-End Engineer Básico:

* HTML (codificar uma estrutura básica)
* CSS (["codar" um layout](http://pt-br.learnlayout.com/), conseguir copiar algum site ou componente simples é um bom indicador do seu nível de conhecimento)
* Lógica de Programação
* JavaScript (inserir um script em uma tela, utilizar libs, criar pequenas coisas)
* Tooling (você vai precisar de algo como o [Webpack](https://webpack.github.io/), [GulpJS](http://gulpjs.com/), [NPM Scripts](https://docs.npmjs.com/misc/scripts))
* Gerenciamento de dependências ([NPM](https://npmjs.org/) e [Yarn](https://yarnpkg.com/pt-BR/))
* O básico de Responsive (aprenda a fazer um layout funcionar em múltiplos dispositivos)
* Git (e saber que [Git](https://git-scm.com/) e [GitHub](http://github.com/) não são a mesma coisa!)

Não corra durante os estudos, pratique muito tudo o que você aprender até ter coragem de falar que sabe aquilo!

Caso você já domine essas coisas, talvez você já seja um(a) **Front-End Engineer Intermediário**!

## Front-End Engineer Intermediário

Requisitos para Front-End Engineer Intermediário:

* HTML (entender sobre Semântica e Acessibilidade)
* Escalabilidade
* Manutenibilidade
* Padrões de Projetos aplicado ao JavaScript e ao CSS (básico, pelo menos)
* Conhece/trabalha com algum Framework CSS e JS
* Pré processadores CSS
* Sabe criar testes unitários
* Entende alguns paradigmas de programação (Ex.: OOP e/ou Funcional)
* Atende aos múltiplos navegadores e suas diferenças
* Criar um ambiente de desenvolvimento ou um Boilerplate para iniciar seus projetos facilmente
* Domina suas ferramentas de trabalho (Editor, S.O., etc)
* Performance (tanto do [JavaScript](https://www.smashingmagazine.com/2012/11/writing-fast-memory-efficient-javascript/), quanto do CSS, [sim do CSS](https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Writing_efficient_CSS))
* SEO (estruturação do HTML, resolver a listinha do PageSpeed, Schema.org)

Se você já domina tudo isso, então você deve estar no nível **Avançado** ou chegando lá.

## Front-End Engineer Avançado

Requisitos para Front-End Engineer Avançado:

* Domina os Paradigmas de Programação e boas práticas com JavaScript
* Entende os problemas e limitações do JavaScript, CSS e HTML
* Consegue criar aplicações 100% com JavaScript e deployar
* Consegue conversar de igual para igual com um Dev Back-End
* Domina os Padrões de Projetos
* Domina pré-processadores e metodologias CSS (como BEM, SMACSS, Atomic ou outras) e boas práticas
* Domina Performance no Front-End
* Entende a importância da Integração Contínua
* Domina testes unitários
* Domina os fluxos de trabalho com Git
* Você entende os custos ($$$) do Front-End (Arquitetura, o que problemas na interface podem gerar negativamente para uma empresa, etc)
* Domina o ambiente do navegador (Render, Painting e etc).
* Entende sobre o protocolo HTTP e o que o cerca ( Segurança, three-way handshake do TCP, como funciona uma requisição)

Pra ver a versão completa, [clica aqui](https://github.com/concretesolutions/front-end-guide).

PS.: Esse documento foi um fork de um outro doc feito pelo @woliveiras pro Training Center.

Espero ter ajudado :+1:

**Comment by ghost** [comment]
@LennonSantos existirão várias linhas de pensamento sobre o assunto e acredito que se você fizer um compilado de tudo será o ideal. Vou fazer minha contribuição com alguns pensamentos que tenho:

* Sempre lembre-se de aprender bem a base (JS, HTML, CSS). Não adianta nada saber do último framework lançado ou tecnologia disruptiva se você não sabe a base.

* Aprenda desde o inicio que testes são importantes.

* Saiba que o seu menor problema será a linguagem de programação. Trabalhar na área de desenvolvimento quer dizer resolver problemas humanos.

* Backend/Frontend não precisa ser isolado e você não precisa temer o termo "fullstack" como se fosse algo grandioso, nada mais é do que um dev web.

* Padrões de projetos são super importantes então não deixe de conhecer e aprender a utiliza-los.

Bom como o @matheusml postou acima tem esse guia que acho que vale a pena você dar uma lida e conferir =) espero ter ajudado \o/

**Comment by felipefialho** [comment]
Fazendo o cross: #415 #75 #132 #222 #353 #506

**Comment by bernardodiasc** [comment]
Vale ressaltar que tem muito mais requerimentos vindo aí para os Front-enders nesses tempos modernos. Front-end developer é o cara que programa pro usuário, e diferentemente de alguns anos atrás, o que vem pela frente é uma **gama ainda maior** de aparelhos e interfaces de usuário para se dar contar e dominar técnicas e tecnologias. HTML e CSS é o arroz com feijão dos browsers, JS é aquela salada (leia churrasco se voce não curte vegetais) maravilhosa com mais de 10 variedades e temperos diferentes.

O Front-end de amanha vai ter que manjar muito além de browsers. Vai ter que programar pra geladeira. Vai ter que programar pra automoveis. Vai ter que programar pra devices que nao possuem interface visual nenhuma. Vai ter que programar pra realidade aumentada e suas variações. Ixi, tem muita coisa vindo por ai, é questão de pouco tempo pra se tornar requerimento, assim como responsividade se tornou requerimento básico desde 2006.

**Comment by MathCunhaDev** [comment]
@matheusml 

Muito obrigado por esse conteúdo, estou começando e estava com muitas dúvidas sobre quando estaria pronto para procurar vagas júnior. Sei que agora tenho que melhorar muita coisa hahaha

Abçs

**Comment by tapmorales** [comment]
@LennonSantos Se me permite. Espero que este texto te acrescente tb. [https://serfrontend.com/blog/como-se-tornar-um-desenvolvedor-frontend/](https://serfrontend.com/blog/como-se-tornar-um-desenvolvedor-frontend/)

Abraço

**Comment by LennonSantos** [comment]
Ótimas respostas, Obrigado show.

**Comment by AmandaFonseca** [comment]
Adorei as dicas meninos acompanhando essa discussão.

@matheusml Estava procurando um trilha de formação de front-end nesse estilo.

**Comment by bbertolani** [comment]
@bernardodiasc não tinha pensando nessas possibilidades. O mercado so tem a crescer...

**Comment by mariojvr** [comment]
Gostei bastante @matheusml . Vou me espelhar...
